src/shared/keyboard-navigation.js

```diff
@@ -0,0 +1,197 @@
+import { useEffect } from 'preact/hooks';
+
+import { ListenerCollection } from './listener-collection';
+
+/**
+ * @param {HTMLElement & { disabled?: boolean }} element
+ */
+function isElementDisabled(element) {
+  return typeof element.disabled === 'boolean' && element.disabled;
+}
+
+/** @param {HTMLElement} element */
+function isElementVisible(element) {
+  return element.offsetParent !== null;
+}
+
+/**
+ * Enable arrow key navigation between interactive descendants of a
+ * container element.
+ *
+ * In addition to moving focus between elements when arrow keys are pressed,
+ * this also implements the "roving tabindex" pattern which sets the `tabindex`
+ * attribute of elements to control which element gets focus when the user
+ * tabs into the container.
+ *
+ * See [1] for a reference of how keyboard navigation should work in web
+ * applications and how it applies to various common widgets.
+ *
+ * @example
+ *   function MyToolbar() {
+ *     const container = useRef();
+ *
+ *     // Enable arrow key navigation between interactive elements in the
+ *     // toolbar container.
+ *     useArrowKeyNavigation(container);
+ *
+ *     return (
+ *       <div ref={container} role="toolbar">
+ *         <button>Bold</bold>
+ *         <button>Italic</bold>
+ *         <a href="https://example.com/help">Help</a>
+ *       </div>
+ *     )
+ *   }
+ *
+ * [1] https://www.w3.org/TR/wai-aria-practices/#keyboard
+ *
+ * @param {import('preact').RefObject<HTMLElement>} containerRef
+ * @param {object} options
+ *   @param {boolean} [options.autofocus] - Whether to focus the first element
```

Comment:
We can discuss at greater length elsewhere, but I'm not a huge fan of this of documenting defaults like this because it can get easily out of sync with the actual default value (unlike the param name and type).